File: challenges/rate-limiter/csharp/RateLimiter.cs
namespace RateLimiterChallenge;

/// <summary>
/// A per-client fixed-window rate limiter.
///
/// IMPLEMENT EVERY MEMBER BELOW. Do not change the public signatures.
/// The test suite in RateLimiterTests.cs defines the exact expected behaviour.
///
/// Example use (illustrative, not a test):
///
///     var limiter = new RateLimiter(maxRequests: 3, windowSeconds: 10);
///     limiter.Allow("alice", 100); // true   (1st in window)
///     limiter.Allow("alice", 101); // true   (2nd)
///     limiter.Allow("alice", 102); // true   (3rd)
///     limiter.Allow("alice", 103); // false  (over the limit, still in window)
///     limiter.Allow("alice", 110); // true   (new window starts)
///
/// Allow always returns a bool. There are no error cases.
/// </summary>
public class RateLimiter
{
    public RateLimiter(int maxRequests, int windowSeconds)
    {
        throw new NotImplementedException();
    }

    public bool Allow(string clientId, long nowEpochSeconds)
    {
        throw new NotImplementedException();
    }
}

File: challenges/rate-limiter/typescript/src/rateLimiter.ts
/**
 * A per-client fixed-window rate limiter.
 *
 * IMPLEMENT EVERY MEMBER BELOW. Do not change the public signatures.
 * The test suite in rateLimiter.test.ts defines the exact expected behaviour.
 *
 * Example use (illustrative, not a test):
 *
 *     const limiter = new RateLimiter(3, 10);
 *     limiter.allow("alice", 100); // true   (1st in window)
 *     limiter.allow("alice", 101); // true
 *     limiter.allow("alice", 102); // true
 *     limiter.allow("alice", 103); // false  (over the limit)
 *     limiter.allow("alice", 110); // true   (new window)
 *
 * `allow` always returns a boolean. There are no error cases.
 */
export class RateLimiter {
  constructor(maxRequests: number, windowSeconds: number) {
    throw new Error("Not implemented");
  }

  allow(clientId: string, nowEpochSeconds: number): boolean {
    throw new Error("Not implemented");
  }
}
